### PR TITLE
Fix cluster health when closing

### PR DESCRIPTION
**Pull request by henningandersen**
When master shuts down it's cluster service, a waiting health request
would fail rather than fail over to a new master.

Fails around once every month in CI:
https://build-stats.elastic.co/goto/49d044995e1b636bfc3cb7e5e2371f3a
For instance:
https://gradle-enterprise.elastic.co/s/r2rocz2i6tah2

### CODE REVIEW

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by henningandersen** [comment]
Failure should be fixed by #62061
@elasticmachine update branch

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterHealthIT.java [hunk 1]
```diff
@@ -309,14 +310,26 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
 
     public void testHealthOnMasterFailover() throws Exception {
         final String node = internalCluster().startDataOnlyNode();
+        boolean withIndex = randomBoolean();
+        if (withIndex) {
+            // create index with many shards to provoke the health request to wait (for green) while master is being shut down.
+            createIndex("test", Settings.builder().put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, randomIntBetween(0, 10)).build());
```
Comment (ywelsch) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterHealthIT.java:316]: I'm confused. How can the cluster ever be green with that many replicas? Should this be `number_of_shards`?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Change looks good. One question on the test

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterHealthIT.java (hunk 1 above)
Comment (henningandersen) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterHealthIT.java:316]: It ensures that the cluster is yellow when the health request is made, making the health request wait on the observer, triggering the call to `onClusterServiceClose` when master is shutdown.

The number of replicas is [cleared to 0](https://github.com/elastic/elasticsearch/pull/61709/files#diff-fb73466f78397039bcf00e8037472624R330) after having fired all the async restarts and done the master restarts. That ensures that all the requests responds with green status.

Comment (ywelsch) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterHealthIT.java:316]: ah ok, can you add a comment to that effect

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Pull request merged**